peas/weather_abstract.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python3
+
+import logging
+import numpy as np
+import re
+import serial
+import sys
+import time
+import requests
+
+from datetime import datetime as dt
+from dateutil.parser import parse as date_parser
+
+import astropy.units as u
+from astropy.units import cds
+from astropy.table import Table
+from astropy.time import Time, TimeISO, TimeDelta
+
+from pocs.utils.messaging import PanMessaging
+
+from . import load_config
+from .PID import PID
+
+def get_mongodb():
+    from pocs.utils.database import PanMongo
+    return PanMongo()
+
+# -----------------------------------------------------------------------------
+#   Base class to read and check weather data
+# -----------------------------------------------------------------------------
+class WeatherAbstract(object):
+
+    """ Base class for checking generic weather data and sending it to the
+        required location.
+    """
+
+    def __init__(self, use_mongo=True):
+        self.config = load_config()
+
+        self.thresholds = self.config.get('thresholds')
+        self.safety_delay = self.config.get('safety_delay', 15.)
+
+        self.db = None
+        if use_mongo:
+            self.db = get_mongodb()
+
+        self.messaging = None
+        self.safe_dict = None
+        self.weather_entries = list()
+
+    def send_message(self, msg, channel='weather'):
+        if self.messaging is None:
+            self.messaging = PanMessaging.create_publisher(6510)
+
+        self.messaging.send_message(channel, msg)
+
+    def capture(self, data):
+        # Make Safety Decision
+        self.safe_dict = self.make_safety_decision(data)
+
+        data['Safe'] = self.safe_dict['Safe']
+        data['Sky condition'] = self.safe_dict['Sky']
+        data['Wind condition'] = self.safe_dict['Wind']
+        data['Gust condition'] = self.safe_dict['Gust']
+        data['Rain condition'] = self.safe_dict['Rain']
+
+        # Store current weather
+        self.weather_entries.append(data)
+
+        if send_message:
+            self.send_message({'data': data}, channel='weather')
+
+        if use_mongo:
+            self.db.insert_current('weather', data)
+
+        return data
+
+    def make_safety_decision(self, current_values):

```

Comment:
This implementation of `make_safety_decision()` makes a lot of assumptions about the types of sensors that the weather station has (i.e. cloud sensors, wind sensors that give average and gust values, and rain sensors), and that's not appropriate for an abstract base class.

All this method should really be doing is calling `_get_status()` to apply the thresholds to the current weather data, then passing the resulting dictionary of statuses to a list of other methods (stored in a `self._safety_methods` member variable, say), ANDing the results to decide whether the overall safety status is 'safe' or 'not safe', and returning the results in a dictionary structure like one you're using.  That's entirely generic and independent of weather station details.

It's then up to the derived classes to specify which methods `make_safety_decision()` should call, and implement those methods using logic specific to their own set of sensors.

Reply:
To facilitate this you should change the individual safety methods so that they all have the same interface.  They should all just take the status dictionary produced by `_get_status()`, extract the statuses that are relevant to their category (e.g. rain, or wind), apply whatever device specific logic is necessary (e.g. combining multiple rain sensors) and return a safe/not safe decision for that category.

Reply:
Have included `statuses = self._get_status()` and then have passed `statuses` as an argument in the `_get_..._safety()` methods.